app/controllers/concerns/enlistment_filters.rb

```diff
@@ -64,7 +67,36 @@ def validate_project
   end
 
   def add_custom_error_msg
-    @project.errors.delete(:description)
-    @project.errors.full_messages.unshift(custom_description_error)
+    project_errors = @project.errors
+    project_errors.delete(:description)
+    project_errors.full_messages.unshift(custom_description_error)
+  end
+
+  def build_code_location
+    CodeLocationBuilder.build do |builder|
+      builder.type = params[:repository][:type]
+      builder.url = params[:repository][:url]
+      builder.repo_params = repository_params
+      builder.code_location_params = code_location_params
+    end
+  end
+
+  def valid_code_location?
+    @code_location = build_code_location
+    @repository = @code_location.repository
+    return render :new, status: :unprocessable_entity unless @code_location.valid?
+  end
+
+  def code_location_exist?
+    code_location = CodeLocation.find_existing(@repository.url, @code_location.module_branch_name)
+    return unless code_location
+    update_repo_username_and_password(code_location)
```

Comment:
If we find an existing Code Location, why is it important to update the username and password? Were we doing this somewhere else or is this needed functionality that was not previously present?

Reply:
We were doing it previously in the controller (here|https://github.com/blackducksoftware/ohloh-ui/blob/master/app/controllers/enlistments_controller.rb#L85). This handles the scenario where a code location is present but its repository's username and password have changed.

Reply:
That's the use case I had assumed.  Very good.